eth/tracers/firehose_concurrency.md

```diff
@@ -0,0 +1,142 @@
+
+1.  **Unit-Level Validation**
+
+    To confirm the functional correctness of the concurrent flushing implementation, a unit test was written that creates and processes 1,000 blocks, flushing each to an `InternalTestingBuffer`. The results were then compared against expected outputs to verify equivalence. The test confirms that the output produced by the concurrent mechanism matches that of the original linear method, thereby validating correctness at the unit level.
+2.  **Integration Testing on Battlefield-Ethereum**
+
+    To verify integration within the battlefield-ethereum environment, the feature was exposed via a new configuration flag: `CONCURRENT_BLOCK_FLUSHING`. This flag determines whether the system uses the default sequential method or the new concurrent implementation. The system can be toggled between these modes with the following commands:
+
+    Original (linear flushing):
+
+    ```bash
+    ./scripts/run_firehose_geth_dev.sh 3.0 prague
+    ```
+
+    Concurrent flushing enabled:
+
+    ```bash
+    CONCURRENT_BLOCK_FLUSHING=true ./scripts/run_firehose_geth_dev.sh 3.0 prague
+    ```
+
+    Behavioral differences were observed through log output. In the concurrent mode, log lines such as:
+
+    ```
+     "Closing channel: flushing the remaining blocks to firehose"
+    ```
+
+    appear when the program is interrupted (e.g., via Ctrl + C), indicating that the concurrent flushing logic and cleanup path are active. These lines are absent in the linear configuration, confirming that the switch is functioning as intended.
+
+    Furthermore, when running the integration test suite using:
+
+    ```bash
+    pnpm test:fh3.0:geth-dev
+    ```
+
+    all tests passed successfully (64 passing), indicating that the concurrent implementation does not introduce regressions in battlefield compatibility.
+3.  **Performance Benchmarking**
+
+    To quantify performance differences, benchmarking was conducted using firehose-ethereum. The following command was used for both the baseline and concurrent configurations:
+
+    ```bash
+    time geth --vmtrace=firehose \
+    --vmtrace.jsonconfig='{"concurrentBlockFlushing":<true|false>}' \
+    --synctarget=0x7ae82cb3e60f13272a59319a4b617022228227258e18e0c5e7404236d773d2a3 \
+    --syncmode=full --holesky --datadir=./geth --db.engine=pebble \
+    --state.scheme=path --port=30305 --authrpc.jwtsecret=jwt.txt \
+    --authrpc.addr=0.0.0.0 --authrpc.port=9551 --authrpc.vhosts="*" \
+    --http --http.addr=0.0.0.0 --http.api=eth,net,web3 --http.port=9545 \
+    --http.vhosts="*" --port=40303 --ws.port=9546 --ipcpath=/tmp/geth.ipc > /dev/null
+    ```
+
+    The benchmark was conducted in two phases:
+
+    * With `concurrentBlockFlushing: false`, the node was synced to block 10,000, the data directory (`./geth`) was removed, and then resynced up to block 100,000.
+    * The same steps were repeated with `concurrentBlockFlushing: true`.
+
```

Comment:
The channel is set to 100 in the code, how was this conducted? Was it done with the channel having 1 or 100 allowed elements?